onmt/dynamic/transform.py

```diff
@@ -0,0 +1,183 @@
+
+    def warm_up(self, vocabs):
+        print("warm up simple transform.")
+
+    def apply(self, src, tgt, **kwargs):
+        return src, tgt
+
+
+class SwitchOutTransform(Transform):
+    def __init__(self, opts):
+        super().__init__(opts)
+
+    def warm_up(self, vocabs):
+        self.vocab = vocabs
+
+    def _swithout(self, tokens):
+        return tokens
+
+    def apply(self, src, tgt, **kwargs):
+        src = self._swithout(src)
+        tgt = self._swithout(tgt)
+        return src, tgt
+
+    def stats(self):
+        pass
+
+
+class SentencePieceTransform(Transform):
+    """A sentencepiece subword transform class."""
+
+    def __init__(self, opts):
+        """Initialize neccessary options for sentencepiece."""
+        super().__init__(opts)
+        self.share_vocab = opts.share_vocab
+        self.src_subword_model = opts.src_subword_model
+        self.tgt_subword_model = opts.tgt_subword_model
+        self.n_samples = opts.n_samples_subword
+        self.theta = opts.theta_subword
+
+    def warm_up(self, vocabs=None):
+        """Load subword models."""
+        import sentencepiece as spm
+        load_src_model = spm.SentencePieceProcessor()
+        load_src_model.Load(self.src_subword_model)
+        if self.share_vocab:
+            self.load_models = {
+                'src': load_src_model,
+                'tgt': load_src_model
+            }
+        else:
+            load_tgt_model = spm.SentencePieceProcessor()
+            load_tgt_model.Load(self.tgt_subword_model)
+            self.load_models = {
+                'src': load_src_model,
+                'tgt': load_tgt_model
+            }
+
+    def _sentencepiece(self, tokens, side='src'):
+        """Do sentencepiece subword encode."""
+        sp_model = self.load_models[side]
+        sentence = ' '.join(tokens)
+        segmented = sp_model.SampleEncodeAsPieces(
+            sentence, nbest_size=self.n_samples, alpha=self.theta)
+        return segmented
+
+    def apply(self, src, tgt, **kwargs):
+        """Apply sentencepiece subword encode to src & tgt."""
+        src = self._sentencepiece(src, 'src')
+        tgt = self._sentencepiece(tgt, 'tgt')
+        return src, tgt
+
+    def __getstate__(self):
+        """Pickling following for rebuild."""
+        return {'opts': self.opts,
+                'share_vocab': self.share_vocab,
+                'src_subword_model': self.src_subword_model,
+                'tgt_subword_model': self.tgt_subword_model,
+                'n_samples': self.n_samples,
+                'theta': self.theta}
+
+    def __setstate__(self, d):
+        """Reload when unpickling from save file."""
+        self.opts = d['opts']
+        self.share_vocab = d['share_vocab']
+        self.src_subword_model = d['src_subword_model']
+        self.tgt_subword_model = d['tgt_subword_model']
+        self.n_samples = d['n_samples']
+        self.theta = d['theta']
+        self.warm_up()
+
+
+AVAILABLE_TRANSFORMS = {
```

Comment:
Maybe put this in a more straightforward place, at the top for instance.

Reply:
It's needed to define all `Transforms` class in order to register them in a dictionary. Will cause `used-before-assignment` in linter if we move it at the top.

Reply:
Makes sense. Maybe it could be split in two files:
- `transform.py` for the transforms themselves;
- `transform_pipe.py` for `TransformPipe` and `TransformStatistics`.

Would this make sense to you?

Also, it would probably be good to have a bit of doc about this as well.
This could for instance take the form of a table:
| Name | Class Name | Description | Compute cost     |
|------|------------|-------------|------------------|
| ...  | ...        | ...         | Low/Medium/Heavy |

Reply:
Apart from making the file smaller, does this split add any other benefits?
If we would like to make it clear, maybe making them in a folder could be a better alternative (like making the base and pipe in __init__, and other specific implementations categorized in different files just like models in fairseq to support user-defined addon)?

Reply:
The folder alternative seems good to me indeed.

Reply:
Move Transforms in the folder and support extensibility done in PR #6.